aten/src/ATen/native/xnnpack/Pooling.cpp

```diff
@@ -0,0 +1,321 @@
+#include <ATen/native/xnnpack/Common.h>
+#include <ATen/native/xnnpack/Factory.h>
+
+namespace at {
+namespace native {
+namespace xnnpack {
+namespace internal {
+namespace max_pool2d {
+
+struct Context final {
+  Operator max_pool_op;
+
+  struct Output final {
+    struct Spatial final {
+      int64_t kernel;
+      int64_t padding;
+      int64_t stride;
+      int64_t dilation;
+      bool ceil_mode;
+    };
+
+    int64_t channels;
+    Spatial height, width;
+  } output;
+
+  static constexpr float kMin = -std::numeric_limits<float>::infinity();
+  static constexpr float kMax = std::numeric_limits<float>::infinity();
+};
+
+namespace {
+
+// Supports NHWC and NCHW FP32 max pooling with any
+//  - kernel size
+//  - padding
+//  - stride
+//  - dilation
+
+bool available(
+    const int64_t input_channels,
+    const int64_t output_channels,
+    const IntArrayRef kernel,
+    const IntArrayRef padding,
+    const IntArrayRef stride,
+    const IntArrayRef dilation,
+    const bool /* ceil_mode */,
+    const float output_min,
+    const float output_max) {
+         // XNNPACK
+  return xnnpack::internal::available() &&
+         // Input
+         (input_channels > 0) &&
+         // Input == Output
+         (input_channels == output_channels) &&
+         // Kernel
+         ((1 <= kernel.size()) && (kernel.size() <= 2)) &&
+         (kernel[Layout::Parameter::height] > 0) &&
+         (kernel[Layout::Parameter::width] > 0) &&
+         // Padding
+         ((1 <= padding.size()) && (padding.size() <= 2)) &&
+         (padding[Layout::Parameter::height] >= 0) &&
+         (padding[Layout::Parameter::width] >= 0) &&
+         // Stride
+         ((1 <= stride.size()) && (stride.size() <= 2)) &&
+         (stride[Layout::Parameter::height] > 0) &&
+         (stride[Layout::Parameter::width] > 0) &&
+         // Dilation
+         ((1 <= dilation.size()) && (dilation.size() <= 2)) &&
+         (dilation[Layout::Parameter::height] > 0) &&
+         (dilation[Layout::Parameter::width] > 0) &&
+         // Output Min / Max
+         (output_max > output_min) &&
+         true;
+}
+
+Context create(
+    const int64_t input_channels,
+    const int64_t output_channels,
+    const IntArrayRef kernel_,
+    const IntArrayRef padding_,
+    const IntArrayRef stride_,
+    const IntArrayRef dilation_,
+    const bool ceil_mode,
+    const float output_min,
+    const float output_max) {
+  // Make sure that parameters are of proper dimension.
+  const auto kernel = expand_param_if_needed(kernel_, "kernel", 2);
+  const auto padding = expand_param_if_needed(padding_, "padding", 2);
+  const auto stride = expand_param_if_needed(stride_, "stride", 2);
+  const auto dilation = expand_param_if_needed(dilation_, "dilation", 2);
+
+  TORCH_CHECK(
+      available(
+          input_channels,
+          output_channels,
+          kernel,
+          padding,
+          stride,
+          dilation,
+          ceil_mode,
```

Comment:
Is `ceil_mode` used somewhere by XNNPACK for maxpool op? PT documentation suggests that it is used to compute output shape, https://pytorch.org/docs/stable/nn.html#torch.nn.MaxPool2d. 
It it is not used then dont we need to figure out what output shape XNNPACK wants and be sure to provide buffer with that shape?
And also impose the restriction on whether maxpool2d can be mapped to xnnpack or not depending on the value of ceil_mode?

Reply:
I wonder if the seg faults in tests are related to the ceil_mode stuff.

Reply:
Right, maybe.  I'll investigate, thanks.